libbpf-tools/runqlat.c

```diff
@@ -0,0 +1,262 @@
+// SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause)
+// Copyright (c) 2020 Wenbo Zhang
+//
+// Based on runqlat(8) from BCC by Bredan Gregg.
+// 10-Aug-2020   Wenbo Zhang   Created this.
+#include <argp.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+#include <unistd.h>
+#include <bpf/libbpf.h>
+#include <bpf/bpf.h>
+#include "runqlat.h"
+#include "runqlat.skel.h"
+#include "trace_helpers.h"
+
+struct env {
+	time_t interval;
+	pid_t pid;
+	int times;
+	bool milliseconds;
+	bool per_process;
+	bool per_thread;
+	bool per_pidns;
+	bool timestamp;
+	bool verbose;
+} env = {
+	.interval = 99999999,
+	.times = 99999999,
+};
+
+static volatile bool exiting;
+
+const char *argp_program_version = "runqlat 0.1";
+const char *argp_program_bug_address = "<bpf@vger.kernel.org>";
+const char argp_program_doc[] =
+"Summarize run queue (scheduler) latency as a histogram.\n"
+"\n"
+"USAGE: runqlat [--help] [-T] [-m] [--pidnss] [-L] [-P] [-p PID] [interval] [count]\n"
+"\n"
+"EXAMPLES:\n"
+"    runqlat         # summarize run queue latency as a histogram\n"
+"    runqlat 1 10    # print 1 second summaries, 10 times\n"
+"    runqlat -mT 1   # 1s summaries, milliseconds, and timestamps\n"
+"    runqlat -P      # show each PID separately\n"
+"    runqlat -p 185  # trace PID 185 only\n";
+
+#define OPT_PIDNSS	1	/* --pidnss */
+
+static const struct argp_option opts[] = {
+	{ "timestamp", 'T', NULL, 0, "Include timestamp on output" },
+	{ "milliseconds", 'm', NULL, 0, "Millisecond histogram" },
+	{ "pidnss", OPT_PIDNSS, NULL, 0, "Print a histogram per PID namespace" },
+	{ "pids", 'P', NULL, 0, "Print a histogram per process ID" },
+	{ "tids", 'L', NULL, 0, "Print a histogram per thread ID" },
+	{ "pid", 'p', "PID", 0, "Trace this PID only" },
+	{ "verbose", 'v', NULL, 0, "Verbose debug output" },
+	{},
+};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	static int pos_args;
+
+	switch (key) {
+	case 'v':
+		env.verbose = true;
+		break;
+	case 'm':
+		env.milliseconds = true;
+		break;
+	case 'p':
+		errno = 0;
+		env.pid = strtol(arg, NULL, 10);
+		if (errno) {
+			fprintf(stderr, "invalid PID: %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	case 'L':
+		if (env.per_pidns || env.per_process) {
+			fprintf(stderr, "pidnss, pids, tids cannot be used "
+				"together\n");
+			argp_usage(state);
+		}
+		env.per_thread = true;
+		break;
+	case 'P':
+		if (env.per_pidns || env.per_thread) {
+			fprintf(stderr, "pidnss, pids, tids cannot be used "
+				"together\n");
+			argp_usage(state);
+		}
+		env.per_process = true;
+		break;
+	case OPT_PIDNSS:
+		if (env.per_process || env.per_thread) {
+			fprintf(stderr, "pidnss, pids, tids cannot be used "
```

Comment:
I'd just do this in one place, after args are parsed, no need to triplicate all this